webapp/hooks/useAllowance.ts

```diff
@@ -1,24 +1,22 @@
 import { useMemo } from 'react'
-import { type Address, type ContractFunctionArgs } from 'viem'
-import { type UseReadContractParameters } from 'wagmi'
+import { type Address, type ContractFunctionArgs, erc20Abi } from 'viem'
+import { type UseReadContractParameters, useReadContract } from 'wagmi'
 
-import { useReadErc20, type Erc20Abi } from './baseErc20'
-
-type AllowanceArgs = ContractFunctionArgs<Erc20Abi, 'view', 'allowance'>
+type AllowanceArgs = ContractFunctionArgs<typeof erc20Abi, 'view', 'allowance'>
 
 type Owner = AllowanceArgs[0]
 type Spender = AllowanceArgs[1]
 
 type Options = {
   args: { owner: Owner; spender: Spender }
-  query?: UseReadContractParameters<Erc20Abi, 'allowance'>['query']
+  query?: UseReadContractParameters<typeof erc20Abi, 'allowance'>['query']
 }
 
 export const useAllowance = (
   erc20Address: Address,
   { args: { owner, spender }, query }: Options,
 ) =>
-  useReadErc20({
+  useReadContract({
```

Comment:
I tested this and the allowance was never returned, thus the code was not working. 

After reviewing it, the issue is that, because the previous implementation used `useReadErc20`, the `abi` was automatically added to the call, which you're not doing in `useReadContract` here. See [here](https://github.com/hemilabs/ui-monorepo/blob/76dddde387188fcb220471ca94b2fe2b2b828821/wagmi-erc20-hooks/baseErc20.ts#L13).

You need to

```tsx
useReadContract({
    address: erc20Address,
    //  Note the abi added here
    abi: erc20Abi,
    args: useMemo(() => [owner, spender], [owner, spender]),
    functionName: 'allowance',
    query,
  })
```

Given we need to change this, I'd also add one extra change:

Given allowance is only for ERC20 tokens, I'd move the checks of `enabled` inside this implementation, as they are all the same for all usages. The function, then would look like this

```tsx
useReadContract({
    address: erc20Address,
    abi: erc20Abi,
    args: useMemo(() => [owner, spender], [owner, spender]),
    functionName: 'allowance',
    query: {
       enabled: !isNativeAddress(erc20Address) && !!owner && !!spender
    },
  })
```

This will simplify the code a bit for all the consumers of this hook.

Reply:
So sorry I couldn't test before pushing, I don't have the token. Thanks for the awesome review, I took it a step further to future proof it Incase we have any other needs for fields in the `query` object. Please test and let me know If I need to make any revision.

```suggestion
  useReadContract({
    abi: erc20Abi,
    address: erc20Address,
    args: useMemo(() => [owner, spender], [owner, spender]),
    functionName: 'allowance',
    query: {
      ...query,
      enabled:
        !isNativeAddress(erc20Address) &&
        !!owner &&
        !!spender &&
        query?.enabled !== false,
    },
  })
```

Reply:
> , I don't have the token.

Did you mean you don't have tokens?

You could get some Sepolia ETH [here](https://cloud.google.com/application/web3/faucet/ethereum/sepolia) and some erc20 tokens [here](https://gho.aave.com/faucet/)

From there, you can tunnel them to Hemi if you need L2 tokens for testing

Reply:
Thank you, I just got some. I'll be able to test before end of day.

Reply:
@gndelia Sorry I'm just getting back to you, had a hectic day. So i tested and it actually didn't work, are there any other factors I'm to consider? It works for non native tokens.

Reply:
@gndelia I misunderstood the criteria, it actually works perfectly 😅. Please test and review.